src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -218,9 +218,11 @@ public void InitializeHeaders()
         {
             if (FrameRequestHeaders == null)
             {
-                RequestHeaders = FrameRequestHeaders = new FrameRequestHeaders();
+                FrameRequestHeaders = new FrameRequestHeaders();
             }
 
+            RequestHeaders = FrameRequestHeaders;
```

Comment:
Yep. Resetting in two separate places was silly to begin with because it can allow for oversights like the one I made here. Just reset them both here.

I don't think there is any need to reset it in ProduceEnd. So I would remove that. I would also update IHttpUpgradeFeature.UpgradeAsync to use FrameResponseHeaders while you're at it.